### PR TITLE
[cloudevents] Add cache for tracking state of cloudevents

### PR DESCRIPTION
**Pull request by waveywaves**
# Changes

- Inject LRU Cache to the context.Context
- Update cache if event is sent.
- Disallow sending duplicate events by checking the cache

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Commit messages includes a project tag in the subject - e.g. [OCI], [hub], [results], [task-loops]

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: cloudevents/config/config-defaults.yaml [hunk 1]
```diff
@@ -21,6 +21,7 @@ metadata:
     app.kubernetes.io/instance: default
     app.kubernetes.io/part-of: tekton-cloudevents
 data:
+  default-cloud-events-sink: http://sockeye.default.10.108.51.172.nip.io/
```
Comment (afrittoli) [inline cloudevents/config/config-defaults.yaml:24]: Please let's not add this default in the repo 🙏

Comment (waveywaves) [reply cloudevents/config/config-defaults.yaml:24]: oh my bad, I forgot I added this, I'l remove it :bow:

File: cloudevents/pkg/reconciler/events/cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+)
+
+// AddEventSentToCache adds the particular object to cache marking it as sent
+func AddEventSentToCache(ctx context.Context, event *cloudevents.Event) error {
+	cacheClient := Get(ctx)
+	if cacheClient == nil {
+		return errors.New("cache client is nil")
+	}
+	cacheClient.Add(event.String(), nil)
```
Comment (afrittoli) [inline cloudevents/pkg/reconciler/events/cache/cache.go:15]: Thanks for adding this. 

I think we'll have to use something different for the key though, as duplicate events might not be exactly the same.
Even with the exact same pipelinerun, timestamps will be different. 

We need to define a function that extracts the part of an event that we consider enough to define two events as the same, and use that as a key.  Probably a combination for event type, resource type and name should be enough to send each event type once for each resource.

In future we might want to send more events, or refresher kind of events, when something changes in the pipeline run but the condition stays the same. We could use the same kind of conditions that are used in the cloud events controller in pipeline today.

Comment (afrittoli) [reply cloudevents/pkg/reconciler/events/cache/cache.go]: Fixed

File: cloudevents/config/config-defaults.yaml (hunk 1 above)
Comment (afrittoli) [reply cloudevents/config/config-defaults.yaml]: fixed

**Review by afrittoli — APPROVED** [review]

**Review by afrittoli — COMMENTED** [review]
/lgtm

**Pull request merged**